Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+        const uint32_t ySuperTileOffset = yCurrent & superTileMask;
+
+        const uint32_t ySuperTile2x2 = ySuperTileOffset >> superTile2x2Shift;
+        const uint32_t ySuperTile2x2Offset = ySuperTileOffset & superTile2x2Mask;
+
+        const uint32_t ySuperTile4x4 = ySuperTile2x2Offset >> superTile4x4Shift;
+        const uint32_t ySuperTile4x4Offset = ySuperTile2x2Offset & superTile4x4Mask;
+
+        const uint32_t ySuperTile8x8 = ySuperTile4x4Offset >> superTile8x8Shift;
+        const uint32_t ySuperTile8x8Offset = ySuperTile4x4Offset & superTile8x8Mask;
+
+        const uint32_t yTile = ySuperTile8x8Offset >> tileShift;
+        const uint32_t yTileOffset = ySuperTile8x8Offset & tileMask;
+
+        const uint32_t tileOffsetRow = ySuperTile * (dw << superTileShift)
+            + ySuperTile2x2 * stride2x2Pixels
+            + ySuperTile4x4 * stride4x4Pixels
+            + ySuperTile8x8 * stride8x8Pixels
+            + (yTile << superTile2x2Shift)
+            + yTileOffset * tileSize;
+
+        const uint32_t* srcRow = src + (sy + y) * spitch + sx;
+        for (uint32_t x = 0; x < sw; x += numberOfPixelsPerBatch) {
+            const uint32_t xCurrent = dx + x;
+            const uint32_t xSuperTile = xCurrent >> superTileShift;
+            const uint32_t xSuperTileOffset = xCurrent & superTileMask;
+
+            const uint32_t xSuperTile2x2 = xSuperTileOffset >> superTile2x2Shift;
+            const uint32_t xSuperTile2x2Offset = xSuperTileOffset & superTile2x2Mask;
+
+            const uint32_t xSuperTile4x4 = xSuperTile2x2Offset >> superTile4x4Shift;
+            const uint32_t xSuperTile4x4Offset = xSuperTile2x2Offset & superTile4x4Mask;
+
+            const uint32_t xSuperTile8x8 = xSuperTile4x4Offset >> superTile8x8Shift;
+            const uint32_t xSuperTile8x8Offset = xSuperTile4x4Offset & superTile8x8Mask;
+
+            const uint32_t xTile = xSuperTile8x8Offset >> tileShift;
+            const uint32_t xTileOffset = xSuperTile8x8Offset & tileMask;
+
+            const uint32_t tileIndex = tileOffsetRow
+                + xTileOffset
+                + xTile * tilePixels
+                + xSuperTile * superTilePixels
+                + xSuperTile2x2 * superTile2x2Pixels
+                + xSuperTile4x4 * superTile4x4Pixels
+                + xSuperTile8x8 * superTile8x8Pixels;
+
+#if HAVE(ARM_NEON_INTRINSICS)
+            ASSERT(tileIndex + 4 <= dw * dh);
+
+            // Prefetch next source and destination memory
+            __builtin_prefetch(srcRow + x + 8, 0, 1);
+            __builtin_prefetch(dst + tileIndex + 8, 1, 1);
+
+            uint32x4_t vdata = vld1q_u32(srcRow + x);
+            vst1q_u32(dst + tileIndex, vdata);
+#else
+            UNUSED_PARAM(dh);
+            dst[tileIndex] = srcRow[x];
+#endif
+        }
+    }
+}
+
+void MemoryMappedGPUBuffer::updateContents(const void* srcData, const IntRect& targetRect, const IntPoint& sourceOffset, unsigned bytesPerLine)
+{
+    ASSERT(m_mappedData);
```

Comment:
I guess this would be very rare, but. Do you think we need to make this a release assertion, or otherwise protect the case where `dstPixels` becomes null in line 468?

Reply:
No longer applicable.